tests/test_models.cc

```diff
@@ -10,247 +10,140 @@
-
-  for (Eigen::Index i = 0; i < joint_predictions.mean.size(); i++) {
-    EXPECT_NEAR(joint_predictions.mean[i], mean_predictions[i], 1e-6);
-    EXPECT_NEAR(joint_predictions.mean[i], marginal_predictions.mean[i], 1e-6);
-    if (joint_predictions.has_covariance()) {
-      EXPECT_NEAR(joint_predictions.covariance(i, i),
-                  marginal_predictions.covariance.diagonal()[i], 1e-6);
-    }
-  }
+Eigen::Index silly_function_to_increment_stack_pointer() {
+  Eigen::VectorXd x(10);
+  return x.size();
 }
 
-/*
- * Here we build two different datasets.  Each dataset consists of targets which
- * have been distorted by non-constant noise (heteroscedastic), we then perform
- * cross-validated evaluation of a GaussianProcess which takes that noise into
- * account, and one which is agnostic of the added noise and assert that taking
- * noise into account improves the model.
- */
-TEST(test_models, test_with_target_distribution) {
-  auto dataset = make_heteroscedastic_toy_linear_data();
-
-  auto folds = leave_one_out(dataset);
-  auto model = MakeGaussianProcess().create();
-  EvaluationMetric<Eigen::VectorXd> rmse =
-      evaluation_metrics::root_mean_square_error;
-  auto scores = cross_validated_scores(rmse, folds, model.get());
-  RegressionDataset<double> dataset_without_variance(dataset.features,
-                                                     dataset.targets.mean);
-  auto folds_without_variance = leave_one_out(dataset_without_variance);
+TYPED_TEST(RegressionModelTester, performs_reasonably_on_linear_data) {
+  auto dataset = this->test_case.get_dataset();
+  auto model = this->test_case.get_model();
 
-  auto scores_without_variance =
-      cross_validated_scores(rmse, folds_without_variance, model.get());
+  const auto fit_model = model.get_fit_model(dataset.features, dataset.targets);
+  const auto pred = fit_model.get_prediction(dataset.features);
+  const auto pred_mean = pred.mean();
 
-  EXPECT_LE(scores.mean(), scores_without_variance.mean());
+  double rmse = sqrt((pred_mean - dataset.targets.mean).norm());
+  EXPECT_LE(rmse, 0.5);
 }
 
-TYPED_TEST(RegressionModelTester, cross_validation_variants) {
-  auto dataset = this->creator.get_dataset();
-  auto folds = leave_one_out(dataset);
-  auto model = this->creator.create();
-  EvaluationMetric<Eigen::VectorXd> rmse =
-      evaluation_metrics::root_mean_square_error;
-  auto cv_scores = cross_validated_scores(rmse, folds, model.get());
+TYPED_TEST(RegressionModelTester, test_predict_variants) {
+  auto dataset = this->test_case.get_dataset();
+  auto model = this->test_case.get_model();
 
-  auto loo_indexers = leave_one_out_indexer(dataset);
-  auto loo_predictions =
-      model->template cross_validated_predictions<MarginalDistribution>(
-          dataset, loo_indexers);
+  const auto fit_model = model.get_fit_model(dataset.features, dataset.targets);
+  silly_function_to_increment_stack_pointer();
```

Comment:
Why is this function needed?

Reply:
Heh, yeah.  That's there to try and catch a tricky bug that resulted from `FitModel` and `Prediction` storing references.  Basically there was a situation where everything worked fine *unless* you called any function in between `get_fit_model` and `get_prediction` at which point the stack pointer would increment and overwrite memory.  The bug was a result of me improperly dealing with references to objects.  At the moment I've switched to just copying everything, but eventually we'll want to be smarter about it and when we do I'd like to make sure we don't hit the same issue.